Convert simulation_ws to ROS2.

timrobotson opened the pull request:
Convert simulation_ws to ROS2. 

Not working yet. Wanted to get PR up for early review / feedback. Need to figure out how to use the new turtlebot_navigation2 as it has completely changed. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

timrobotson committed:
Convert simulation_ws to ROS2

- Convert package.xml CMakeLists, launch files to ROS2

timrobotson committed:
Remove CMakeLists, build with pure python

timrobotson committed:
Update turtlebot3_description_reduced_mesh to ROS2

timrobotson committed:
Fix CMakeLists and packaging files

timrobotson committed:
Update README, fix bugs

- Update README with ros2 relevant instructions.
- Use os.path.join for all Includes
- Fix bugs in launch files

timrobotson committed:
Add turtlebot3_description_reduced_mesh from object-tracker

ryanewel commented on simulation_ws/src/aws_robomaker_simulation_common/src/aws_robomaker_simulation_common/route_manager.py:
```diff
@@ -55,57 +60,62 @@ class RouteManager():
+            route_file_contents = f.read()
+        route_yaml = yaml.safe_load(route_file_contents)
+
+        self.route_mode = route_yaml['mode']
```
I think the mode should be left as a parameter. The way this worked in ROS1 you could use the same route file and just overwrite the mode parameter to do an in-order traversal of the poses, but this implementation cuts out that functionality.

timrobotson replied:
Oh yea true. Can fix once I get route_manager.py working in a later PR.

ryanewel commented on simulation_ws/src/cloudwatch_simulation/launch/bookstore.launch:
```diff
@@ -20,5 +20,4 @@
     <arg name="y_pos" default="5.5"/>
     <arg name="z_pos" default="0.3"/>
   </include>
- <include file="$(find cloudwatch_simulation)/launch/spawn_balls.launch"/>
```
Why aren't we spawning the balls anymore?

timrobotson replied:
This launch file isn't in the repo, so it never worked. I think the balls were just added to the world instead (https://github.com/aws-robotics/aws-robomaker-sample-application-cloudwatch/blob/master/simulation_ws/src/cloudwatch_simulation/worlds/empty.world#L45) and this was never removed.

ryanewel commented on simulation_ws/src/cloudwatch_simulation/launch/turtlebot3_navigation.launch.py:
```diff
@@ -0,0 +1,77 @@
+        ),
+        launch.actions.DeclareLaunchArgument(
+            name='open_rviz',
+            default_value='true'
```
I feel like the default to this should be false since we set it to that anyway.

timrobotson replied:
This was taken from https://github.com/aws-robotics/aws-robomaker-small-house-world/blob/master/launch/turtlebot3_navigation.launch where it was set to true previously. Wanted to keep it consistent with ROS1.

ryanewel reviewed (changes requested):
Could you add the simulation ws to the .travis.yml? It's not currently being built

timrobotson committed:
Fix bookstore simulation

- Add navigation_msgs for turtlebot3_navigation file
- Fix simulation_common package to install files into correct locatino
- Make empty_world work correctly
- Bookstore_navigation loads robot and bookstore, but navigation isn't
happening yet.
- Port turtlebot_description_reduced_mesh changes from hello-world
sample app

timrobotson commented:
empty_world is now working and turtlebot_navigation works and loads the environment + robot but the robot isn't moving yet. The code still needs a little more cleanup too.

AAlon committed:
Removing CEs from robot_ws/.rosinstall and using apt instead

AAlon committed:
Add simulation_ws build

timrobotson committed:
Clean up world / turtlebot spawning

- Pass the gazebo model path to the bookstore world launch file so it's
not hardcoded any more.
- Change spawning turtlebot to use spawn_entity.py now instead of
running shell commands.
- Add ability to set turtlebot initial pose
- Convert xacro files to urdf because it can't be done on the fly with
spawn_entity.py

timrobotson commented:
- Added ability to set the turtlebot spawn location and launching it via spawn_entity.py file now. 
- Passing the gazebo_model_path to the bookstore-world launch file instead of hard coding it into that file (requires https://github.com/aws-robotics/aws-robomaker-bookstore-world/pull/5 to be merged)

ryanewel commented on simulation_ws/src/cloudwatch_simulation/launch/bookstore_turtlebot3_navigation.launch.py:
```diff
@@ -0,0 +1,89 @@
+            ),
+            launch_arguments={
+                'gui': launch.substitutions.LaunchConfiguration('gui'),
+                'gazebo_model_path': os.path.split(get_package_share_directory('turtlebot3_description_reduced_mesh'))[0],
```
I'd prefer `os.path.dirname` here. It gives the same result and makes the intention clearer. Same in other places

timrobotson replied:
I think dirname would be less clear as I think people would assume `os.path.dirname(get_package_share_directory('turtlebot3_description_reduced_mesh'))` would have `share` on the end (as share is a directory) when it actually strips the share directory. I think the [0] is more clear as people realize it's omitting the last bit of the path. They do both do the exact same thing:

```
In [2]: os.path.split('/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world/share')[0]                                        
Out[2]: '/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world'
In [3]: os.path.dirname('/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world/share')                                         
Out[3]: '/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world'
In [5]: os.path.dirname('/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world/share/')                                        
Out[5]: '/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world/share'
In [6]: os.path.split('/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world/share/')[0]                                       
Out[6]: '/opt/workspace/simulation_ws/install/aws_robomaker_bookstore_world/share'
```

The function `get_package_share_directory()` returns the path without a slash on the end.
